### ISSUE TITLE
htex caching return values unexpectedly, which may cause problems with mutable return values

### ISSUE THREAD

**Issue by benclifford** [Parsl/parsl#2555]
**Describe the bug**
In some circumstances, htex with return the same object from multiple invocations of an app. This is undesirable when return objects are mutable (for example lists). This behaviour would be acceptable when memo/app caching is enabled, but not otherwise.

It looks like this happens because deserialize is wrapped with lru_cache in parsl.serialize.base:

```
        self.deserialize = functools.lru_cache(maxsize=maxsize)(self.deserialize)
```

Removing that line makes this test case work with htex.

This test case does not fail with thread local or the WQ executor.

This was found by Sander Vandenhaute.

**To Reproduce**
Run this pytest case with `--config parsl/tests/configs/htex_local.py`

```
@parsl.python_app
def return_range(x):
    return list(range(x))

@pytest.mark.skip("this demonstrates an issue, by failing")
def test_range_identities():
    # these don't necessarily make sense with caching turned on so beware.
    x = 3

    fut1 = return_range(x)
    res1 = fut1.result()

    fut2 = return_range(x)
    res2 = fut2.result()

    assert(fut1 != fut2)
    assert(id(fut1) != id(fut2))
    assert(id(res1) != id(res2))
```

Steps to reproduce the behavior, for e.g:
1. Setup Parsl 0.8.0 with Python 3.6 on cluster
2. Run a test script
3. Wait 5 mins

**Expected behavior**
Without caching/memoisation turned on, python objects returned by different tasks should be distinct.

**Comment by mcourteaux** [comment]
Could by any chance this help? I was following the path of submitting a job and came out here:

https://github.com/Parsl/parsl/blob/7a71fc516fae91626c30f651a3bb5c4cc0593fba/parsl/dataflow/dflow.py#L610-L615

I think one should return a deepcopy of the cached result. Me, not knowing __anything__ about this project, would think -- as a general guideline from CS -- that returning a non-const reference to a mutable object stored in an in-memory cache, is wrong.

Eg:

```py

import copy

 memo_fu = self.memoizer.check_memo(task_record) 
 if memo_fu: 
     logger.info("Reusing cached result for task {}".format(task_id)) 
     task_record['from_memo'] = True 
     assert isinstance(memo_fu, Future) 
     return copy.deepcopy(memo_fu) # returning from a cache should never allow the cache to be modified.

```

**Comment by benclifford** [comment]
@mcourteaux the bit I was looking at was elsewhere, where the result deserializer is wrapped in a lru_cache (sometimes) and performs caching without user opt-in.

I think it's wrong in general to be doing that - my gut feeling is that it would be better to not cache deserialization at all, rather than paying the cost of caching and then duplicating the object afterwards anyway with deepcopy.

The end effect would mostly be the same.

But, the bit of code you pasted is only when a user has chosen to turn on memoization for their parsl functions - it wouldn't make a deep copy when the default of non-memoization is happening. That deepcopy would also only fire on the 2nd and subsequent returns of values, so if a user mutates the result from the first time it comes back, then all subsequent deep copies will be of that mutated result. You'd have to, I think, *always* copy the result object when memoization was turned on, even if it turns out that result object is only used once.

For the cases where a user *has* chosen to turn on memoization, I'm happy for parsl to impose stricter rules about what you're allowed to do with the result: you *chose* to share results. Dealing with sharing mutable objects is a regular part of Python - we aren't programming a pure functional language.

**Comment by benclifford** [comment]
@yadudoc do you have any opinions on removing the lru_cache decorator from the deserialize in this code? It's dangerous when deserializing mutable objects.

An alternative is to force this cached-deserialization only on callables, which is less dangerous (but not entirely mutability free) by implementing a second serialisation type (rather than sharing the same serialization methods between "code" and "data). I would prefer to avoid that complexity without performance evidence that it's worth it.

**Comment by benclifford** [comment]
fixed by #2794
